swarm/swap_test.go

```diff
@@ -0,0 +1,468 @@
+			}
+			log.Trace("file uploaded", "node", id, "key", key.String())
+			files = append(files, file{
+				addr:   key,
+				data:   data,
+				nodeID: id,
+			})
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
+			//we use a special retrieve function for swap which is optimized for parallel requests
+			//but does not leave many cascaded requests floating around
+			if retrieveForSwap(sim, files) == 0 {
+				break
+			}
+		}
+
+		//iterate all nodes
+		for _, node := range sim.NodeIDs() {
+			item, ok := sim.NodeItem(node, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+
+			//submap for each node is a map of all nodes with the balance for that node
+			subBalances := make(map[enode.ID]int64)
+
+			//iterate all nodes again...
+			//get all balances with other peers for every node
+			for _, n := range sim.NodeIDs() {
+				if node == n {
+					continue
+				}
+
+				//get the peer's balance with this node
+				balance, err := swarm.swap.GetPeerBalance(n)
+				if err == nil {
+					subBalances[n] = balance
+					log.Debug(fmt.Sprintf("Balance of node %s to node %s: %d", node.TerminalString(), n.TerminalString(), balance))
+				} else {
+					log.Debug(fmt.Sprintf("Node %s has no balance with node %s", node.TerminalString(), n.TerminalString()))
+				}
+			}
+			//update the map for this node
+			balancesMap[node] = subBalances
+		}
+
+		return nil
+	})
+
+	sim.Close()
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	//print all the balances if requested
+	if *printStats {
+		for k, v := range balancesMap {
+			fmt.Println(fmt.Sprintf("node %s balances:", k.TerminalString()))
+			for kk, vv := range v {
+				fmt.Println(fmt.Sprintf(".........with node %s: balance %d", kk.TerminalString(), vv))
+			}
+		}
+	}
+
+	//now iterate the whole map
+	//and check that every node k has the same
+	//balance with a peer as that peer with the node,
+	//but in inverted signs
+
+	//iterate the map
+	success := true
+	for k, mapForK := range balancesMap {
+		//iterate the submap
+		for n, balanceKwithN := range mapForK {
+			//iterate the main map again
+			mapForSubK := balancesMap[n]
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", k.TerminalString(), n.TerminalString(), balanceKwithN))
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", n.TerminalString(), k.TerminalString(), mapForSubK[k]))
+			//...check that they have the same balance in Abs terms and that it is not 0
+			if balanceKwithN+mapForSubK[k] != 0 && balanceKwithN != 0 {
+				log.Error(fmt.Sprintf("Expected balances to be a+b = 0 AND balance(a) != 0, but they are not, balance k with n:  %d, balance n with k: %d", balanceKwithN, mapForSubK[k]))
+				success = false
+			}
+		}
+	}
+	if !success {
+		t.Fatal("Expected balances to be symmetrical, but they were not")
+	}
+	log.Debug("test terminated")
```

Comment:
not necessary IMHO

Reply:
It's a debug log, and when debugging, it may be helpful to identify where you are in the test. (Note that this denotes the actual test finished, then there is some cleanup unrelated to this test, so it may be informative).

Reply:
Because the test is run by the `sim.Run()` method, which, after the code you see here, will do some cleanup and finish. So it is helpful to see that the actual test code terminated, and if there is still an error, that it may be in the simulation package.